### PR TITLE
Gin wrapper

### PR DESCRIPTION
**Pull request by enoodle**
Adding support for gin wrapper

gin is a webserver framework for golang (like flask is for python)

In future PR we will add the creation of the trigger event.

### CODE REVIEW

File: wrappers/gin/gin_test.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
+package epsagongin
+
+import (
+	"fmt" // EREZ
```
Comment (maorlx) [inline wrappers/gin/gin_test.go:4]: remove comment

File: epsagon/generic_wrapper.go [hunk 2]
```diff
@@ -172,8 +184,8 @@ func GoWrapper(config *Config, wrappedFunction interface{}, args ...string) Gene
 	}
 }
 
-// GoWrapper wraps the function with epsagon's tracer
-func ConcurrentGoWrapper(config *Config, wrappedFunction interface{}, args ...string) GenericFunction {
+// injectContextGoWrapper wraps the function with epsagon's tracer
+func injectContextGoWrapper(config *Config, wrappedFunction interface{}, args ...string) GenericFunction {
```
Comment (maorlx) [inline epsagon/generic_wrapper.go:188]: How would we keep the option to use the ConcurrentGoWrapper now? (I see this function is not used)

File: wrappers/gin/gin.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			config = &epsagon.Config{}
+		}
+		wrapperTracer := tracer.CreateTracer(&config.Config)
+		wrapperTracer.Start()
+		defer wrapperTracer.Stop()
+
+		if c.Keys == nil {
+			c.Keys = make(map[string]interface{})
+		}
+		c.Keys["EpsagonTracer"] = wrapperTracer
```
Comment (maorlx) [inline wrappers/gin/gin.go:36]: use TracerKey

Comment (enoodle) [reply wrappers/gin/gin.go:36]: :+1:

File: epsagon/generic_wrapper.go (hunk 2 above)
Comment (enoodle) [reply epsagon/generic_wrapper.go:188]: This is a mistake, will fix

File: wrappers/gin/gin_test.go (hunk 1 above)
Comment (enoodle) [reply wrappers/gin/gin_test.go:4]: :+1:

File: wrappers/gin/gin.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
+package epsagongin
```
Comment (maorlx) [inline wrappers/gin/gin.go:1]: should update the README with this new support

File: wrappers/gin/gin.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+		if c.Keys == nil {
+			c.Keys = make(map[string]interface{})
+		}
+		c.Keys[TracerKey] = wrapperTracer
+		wrapper := epsagon.WrapGenericFunction(
+			handler,
+			config,
+			wrapperTracer,
+			false,
```
Comment (maorlx) [inline wrappers/gin/gin.go:41]: If the wrapped handler will wrap internal requests events - they obviously won't be collected. I guess we should document it somewhere (probably in the README)

File: wrappers/gin/gin.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+			fmt.Sprintf("%s%s", hostname, relativePath),
+		)
+		wrapper.Call(c)
+	}
+}
+
+// Handle is a wrapper for gin.RouterGroup.Handle that adds epsagon instrumentaiton and event triggers
+// to all invocations of that handler
+func (router *GinRouterWrapper) Handle(httpMethod, relativePath string, handlers ...gin.HandlerFunc) gin.IRoutes {
+	if len(handlers) >= 1 {
```
Comment (maorlx) [inline wrappers/gin/gin.go:51]: would add some Epsagon debug here

File: wrappers/gin/gin.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+		}
+		wrapperTracer := tracer.CreateTracer(&config.Config)
+		wrapperTracer.Start()
+		defer wrapperTracer.Stop()
+
+		if c.Keys == nil {
+			c.Keys = make(map[string]interface{})
+		}
+		c.Keys[TracerKey] = wrapperTracer
+		wrapper := epsagon.WrapGenericFunction(
```
Comment (maorlx) [inline wrappers/gin/gin.go:37]: I wonder whether we want to wrap the customer handler with some code which will extract the request method (and maybe more data from the given gin context?)

**Review by maorlx — COMMENTED** [review]
looks good :)

**Pull request merged**